### PR TITLE
RSTcheckBear: Allow ignoring roles and directives

### PR DESCRIPTION
**Pull request by mridubhatnagar**
Closes https://github.com/coala/coala-bears/issues/1723

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/rest/RSTcheckBear.py [hunk 1]
```diff
@@ -1,8 +1,19 @@
+import re
 from coalib.bearlib.abstractions.Linter import linter
```
Comment (jayvdb) [inline bears/rest/RSTcheckBear.py]: add a blank line between standard libraries (re) and other imports

File: bears/rest/RSTcheckBear.py [hunk 2]
```diff
@@ -32,16 +43,40 @@ class RSTcheckBear:
...
+             Comma seperated list of roles to be ignored
         """
+
         args = ()
-        if code_block_language_ignore:
-            args = ('--ignore-language=' +
-                    ','.join(code_block_language_ignore),)
-        return args + (filename,)
+        args += ((
+            '--ignore-language=' + ','
```
Comment (jayvdb) [inline bears/rest/RSTcheckBear.py]: you do not need to modify the lines regarding `--ignore-language=`

**Comment by supergr35** [comment]
We highly recommend running `coala` on your own code locally so that you can solve the gitmate errors.

**Comment by jayvdb** [comment]
Missing coverage  10-16, 68, 67->68

File: bears/rest/RSTcheckBear.py [hunk 3]
```diff
@@ -32,16 +45,44 @@ class RSTcheckBear:
...
             ``python``, ``rst``.
+
+        :param ignore_directives:
+             Comma sepereated list of directives to be ignored
+
+        :param ignore_roles:
+             Comma seperated list of roles to be ignored
         """
+        content = ''.join(file)
+        error = check(content)
```
Comment (jayvdb) [inline bears/rest/RSTcheckBear.py]: This method, called `create_arguments`, does not do any checks.

Comment (mridubhatnagar) [inline bears/rest/RSTcheckBear.py]: From rstcheck I had imported check. Check tells the errors if there are any.

**Comment by mridubhatnagar** [comment]
@jayvdb can you please review the changes. I am still facing issues with the test cases. Though have addressed the previous changes you suggested.

File: bears/rest/RSTcheckBear.py [hunk 4]
```diff
@@ -1,8 +1,20 @@
...
 from dependency_management.requirements.PipRequirement import PipRequirement
 from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
 
 
+def _ignore_unknown(errors):
+    for error in errors:
+        msg = error.full_message
+        res = re.search(
+            r'No directive entry for "[\w|\-]+"|'
+            r'No role entry for "[\w|\-]+"', msg)
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: you can shorten this to
```
No (?:directive|role) entry for "[\w|\-]+"
```

File: bears/rest/RSTcheckBear.py [hunk 5]
```diff
@@ -32,16 +44,39 @@ class RSTcheckBear:
...
+            '--ignore-directives=' + ','
+            .join(ignore_directives),)
+            if ignore_directives else ()
+        )
+
+        args += ((
+            '--ignore-roles=' + ','
+            .join(ignore_roles),)
+            if ignore_roles else ()
+        )
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: maybe use a list for `args` instead, and append stuff when conditions are satisfied

File: bears/rest/RSTcheckBear.py [hunk 6]
```diff
@@ -1,8 +1,20 @@
+import re
+
 from coalib.bearlib.abstractions.Linter import linter
 from dependency_management.requirements.PipRequirement import PipRequirement
 from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
 
 
+def _ignore_unknown(errors):
```
Comment (Makman2) [inline bears/rest/RSTcheckBear.py]: where is this function used actually?

File: tests/rest/RSTcheckBearTest.py [hunk 7]
```diff
@@ -7,6 +10,22 @@
...
+
+def test_ignore_unknown(errors):
+    for error in errors:
+        msg = error.full_message
+    unittest.assertRegex('[\w|\-]+', re)
+    unittest.assertEqual(
+        msg,
+        'No directive entry for "argparse" in module'
+        '"docutils.parsers.rst.languages.en"'
+    )
```
Comment (Makman2) [inline tests/rest/RSTcheckBearTest.py]: tests need to be defined inside a test class ;) That's why you don't get 100% coverage, as this function never gets picked up. But do you need `_ignore_unknown` anyway? It's never used

File: bears/rest/RSTcheckBear.py (hunk 6 above)
Comment (mridubhatnagar) [inline bears/rest/RSTcheckBear.py]: @Makman2 earlier I was calling this function from inside create arguments. But that seemed to be a wrong way. So, removed it .

File: bears/rest/RSTcheckBear.py [hunk 8]
```diff
@@ -32,16 +44,39 @@ class RSTcheckBear:
...
+                         ignore_roles: list=()):
         """
         :param code_block_language_ignore:
             Comma seperated list for languages of which code blocks
             should be ignored. Code block of following languages
             can be detected: ``bash``, ``c``, ``cpp``, ``json``,
             ``python``, ``rst``.
+
+        :param ignore_directives:
+             Comma sepereated list of directives to be ignored
```
Comment (Asnelchristian) [inline bears/rest/RSTcheckBear.py]: missing dot `.`

File: bears/rest/RSTcheckBear.py [hunk 9]
```diff
@@ -32,16 +44,39 @@ class RSTcheckBear:
...
+
+        :param ignore_roles:
+             Comma seperated list of roles to be ignored
```
Comment (Asnelchristian) [inline bears/rest/RSTcheckBear.py]: same here ;)

File: bears/rest/RSTcheckBear.py [hunk 10]
```diff
@@ -32,16 +44,39 @@ class RSTcheckBear:
...
         """
+
```
Comment (Asnelchristian) [inline bears/rest/RSTcheckBear.py]: we usually don't add a blank line here (after the settings' docs)

Comment (Asnelchristian) [inline bears/rest/RSTcheckBear.py]: but if you want to (which i don't recommend for consistency ) do it in another commit

**Comment by umeshksingla** [comment]
I can help with this? I actually need this feature badly, otherwise, I need to turn off linting on many files with a custom directive for my project.

**Comment by Makman2** [comment]
@umeshksingla sure :) If you want to take it over, just get the commits and do the modifications that the reviewers left (or add other ones you think are better/more appropriate/etc).
-->
```bash
git fetch origin pull/1757/head:rst-ignore-roles-and-directives && git checkout rst-ignore-roles-and-directives
```
If you open a PR, I would close this one then :+1:

**Comment by umeshksingla** [comment]
@Makman2 don't close this yet, I'll tell you soon. And thanks for the ready-to-go pointer.

File: bears/rest/RSTcheckBear.py (hunk 6 above)
Comment (umeshksingla) [inline bears/rest/RSTcheckBear.py]: @Makman2 why do we need such a function anyway? I don't understand the intention behind this function and on top of that, the regex it checks for is very weird. 
Do we even use `unittest.assertRegex`kind things in coala-bears?
Also, I might need your help with tests later.

**Comment by umeshksingla** [comment]
@Makman2 Yes, please close this one.

**Comment by Makman2** [comment]
@umeshksingla please open a PR first :+1: :smile:

**Comment by umeshksingla** [comment]
@Makman2 here, PR #2018.

**Comment by Makman2** [comment]
#2018 is taking over.
